File: platforms/android/src/main/kotlin/dev/nse/NSE.kt
package dev.nse

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import androidx.biometric.BiometricPrompt
import java.security.KeyPairGenerator
import java.security.KeyStore

/**
 * Nostr Secure Enclave — Android implementation
 * Hardware-backed key management using StrongBox/TEE P-256 to protect secp256k1 keys
 */
object NSE {

    private const val KEYSTORE_PROVIDER = "AndroidKeyStore"
    private const val ENCLAVE_KEY_ALIAS = "dev.nse.enclave.p256"
    private const val BLOB_PREF_KEY = "dev.nse.encrypted.secp256k1"

    /**
     * Generate a new secp256k1 keypair, protected by a StrongBox/TEE P-256 key
     */
    suspend fun generate(context: Context): KeyInfo {
        // TODO: Phase 2 implementation
        // 1. Generate P-256 key in StrongBox/TEE (setIsStrongBoxBacked, biometric-gated)
        // 2. Generate secp256k1 keypair
        // 3. Derive AES-256-GCM key from P-256 via ECDH + HKDF
        // 4. Encrypt secp256k1 private key with AES key
        // 5. Store encrypted blob in EncryptedSharedPreferences
        // 6. Zero the plaintext: privateKeyBytes.fill(0)
        // 7. Return pubkey + npub
        throw NotImplementedError("Not yet implemented")
    }

    /**
     * Sign a Nostr event
     * Biometric unlock → decrypt secp256k1 key → Schnorr sign → zero memory
     */
    suspend fun sign(event: NostrEvent, promptInfo: BiometricPrompt.PromptInfo): SignedEvent {
        // TODO: Phase 2 implementation
        // 1. Biometric unlock via BiometricPrompt
        // 2. Access KeyStore P-256 key
        // 3. Derive AES key
        // 4. Decrypt secp256k1 private key into memory
        // 5. Schnorr sign the event (BIP-340)
        // 6. Zero the plaintext: privateKeyBytes.fill(0)
        // 7. Return signed event
        throw NotImplementedError("Not yet implemented")
    }

    /** Get the hex pubkey (does not require biometric unlock) */
    suspend fun getPublicKey(context: Context): String {
        throw NotImplementedError("Not yet implemented")
    }

    /** Get the bech32 npub */
    suspend fun getNpub(context: Context): String {
        throw NotImplementedError("Not yet implemented")
    }

    /** Check if a key exists */
    fun exists(): Boolean {
        val keyStore = KeyStore.getInstance(KEYSTORE_PROVIDER)
        keyStore.load(null)
        return keyStore.containsAlias(ENCLAVE_KEY_ALIAS)
    }

    /** Wipe all key material */
    suspend fun destroy(context: Context) {
        // TODO: Delete KeyStore entry + EncryptedSharedPreferences blob
        val keyStore = KeyStore.getInstance(KEYSTORE_PROVIDER)
        keyStore.load(null)
        keyStore.deleteEntry(ENCLAVE_KEY_ALIAS)
    }

    // Types

    data class KeyInfo(
        val pubkey: String,
        val npub: String,
        val createdAt: Long,
        val hardwareBacked: Boolean
    )

    data class NostrEvent(
        val kind: Int,
        val content: String,
        val tags: List<List<String>>,
        val createdAt: Long
    )

    data class SignedEvent(
        val id: String,
        val pubkey: String,
        val sig: String,
        val kind: Int,
        val content: String,
        val tags: List<List<String>>,
        val createdAt: Long
    )
}

File: platforms/ios/Sources/NSE/NSE.swift
import Foundation
import CryptoKit
import LocalAuthentication
import secp256k1

/// Nostr Secure Enclave — iOS implementation
/// Hardware-backed key management using Secure Enclave P-256 to protect secp256k1 keys
public final class NSE {

    /// Tag used to identify the Secure Enclave P-256 key in Keychain
    private static let enclaveKeyTag = "dev.nse.enclave.p256"

    /// Tag used to identify the encrypted secp256k1 blob in Keychain
    private static let blobKeyTag = "dev.nse.encrypted.secp256k1"

    // MARK: - Public API

    /// Generate a new secp256k1 keypair, protected by a Secure Enclave P-256 key
    public static func generate() async throws -> KeyInfo {
        // TODO: Phase 1 implementation
        // 1. Generate P-256 key in Secure Enclave (non-exportable, biometric-gated)
        // 2. Generate secp256k1 keypair
        // 3. Derive AES-256-GCM key from P-256 via ECDH + HKDF
        // 4. Encrypt secp256k1 private key with AES key
        // 5. Store encrypted blob in Keychain
        // 6. Zero the plaintext secp256k1 key from memory
        // 7. Return pubkey + npub
        fatalError("Not yet implemented")
    }

    /// Sign a Nostr event
    /// Biometric unlock → decrypt secp256k1 key → Schnorr sign → zero memory
    public static func sign(_ event: NostrEvent) async throws -> SignedEvent {
        // TODO: Phase 1 implementation
        // 1. Biometric unlock via LAContext
        // 2. Access Secure Enclave P-256 key
        // 3. Derive AES key
        // 4. Decrypt secp256k1 private key into memory
        // 5. Schnorr sign the event (BIP-340)
        // 6. Zero the plaintext key: privateKeyData.resetBytes(in: 0..<privateKeyData.count)
        // 7. Return signed event
        fatalError("Not yet implemented")
    }

    /// Get the hex pubkey (does not require biometric unlock)
    public static func getPublicKey() async throws -> String {
        // TODO: Read from Keychain metadata (pubkey stored alongside encrypted blob)
        fatalError("Not yet implemented")
    }

    /// Get the bech32 npub
    public static func getNpub() async throws -> String {
        // TODO: Derive from getPublicKey() using bech32 encoding
        fatalError("Not yet implemented")
    }

    /// Check if a key exists in Keychain
    public static func exists() -> Bool {
        // TODO: Query Keychain for blobKeyTag
        return false
    }

    /// Wipe all key material (Secure Enclave key + encrypted blob)
    public static func destroy() async throws {
        // TODO: Delete both Keychain items
        // 1. Delete encrypted secp256k1 blob
        // 2. Delete Secure Enclave P-256 key
    }

    // MARK: - Types

    public struct KeyInfo {
        public let pubkey: String
        public let npub: String
        public let createdAt: Date
        public let hardwareBacked: Bool
    }

    public struct NostrEvent {
        public let kind: Int
        public let content: String
        public let tags: [[String]]
        public let createdAt: Int

        public init(kind: Int, content: String, tags: [[String]], createdAt: Int) {
            self.kind = kind
            self.content = content
            self.tags = tags
            self.createdAt = createdAt
        }
    }

    public struct SignedEvent {
        public let id: String
        public let pubkey: String
        public let sig: String
        public let kind: Int
        public let content: String
        public let tags: [[String]]
        public let createdAt: Int
    }
}
